src/generic_datatype.cpp

```diff
@@ -0,0 +1,101 @@
+#include "generic_datatype.h"
+
+#include <memory>
+#include <unordered_map>
+
+#include "assert.h"
+
+using namespace std;
+
+namespace smt {
+GenericDatatypeConstructorDecl::GenericDatatypeConstructorDecl(
+    const std::string & name)
+    : cons_name(name)
+{
+}
+
+void GenericDatatypeConstructorDecl::add_new_selector(
+    const shared_ptr<selectorComponents> & newSelector)
+{
+  selector_vector.push_back(*newSelector);
+}
+
+std::vector<selectorComponents>
+GenericDatatypeConstructorDecl::get_selector_vector()
+{
+  return selector_vector;
+}
+
+std::string GenericDatatypeConstructorDecl::get_name() const
+{
+  return cons_name;
+}
+
+int GenericDatatypeConstructorDecl::get_selector_count() const
+{
+  return selector_vector.size();
+}
+bool GenericDatatypeConstructorDecl::compare(
+    const DatatypeConstructorDecl & d) const
+{
+  // Why won't type casting like this work?
+  // GenericDatatypeConstructorDecl gdtc = (GenericDatatypeConstructorDecl) d;
+  return cons_name == static_pointer_cast<GenericDatatypeConstructorDecl>(d)->get_name();
+}
+
+GenericDatatype::GenericDatatype(const DatatypeDecl & dt_declaration,
+                                 const std::string & s)
+    : dt_decl(dt_declaration), name(s)
+{
+}
+
+void GenericDatatype::add_constructor(
+    const DatatypeConstructorDecl & dt_cons_decl)
+{
+  cons_decl_vector.push_back(dt_cons_decl);
+}
+
+void GenericDatatype::add_selector(
+    const DatatypeConstructorDecl & dt_cons_decl,
+    const std::shared_ptr<selectorComponents> & newSelector)
+{
+  for (unsigned int i = 0; i < cons_decl_vector.size(); ++i)
+  {
+    if (cons_decl_vector[i] == dt_cons_decl)
+    {
+      static_pointer_cast<GenericDatatypeConstructorDecl>(cons_decl_vector[i])->get_selector_vector().push_back(*newSelector);
```

Comment:
just call this function:
https://github.com/afnaanhashmi/smt-switch/blob/524b305eedff7cd3d3ed08c7542a9c720404deeb/include/generic_datatype.h#L35

And in the implementation of that function, throw an exception if the selector already exists.